### DISCUSSION TITLE
Handling of translations

### DISCUSSION THREAD

**Discussion by leamas** [opencpn-manuals/main#32] [General]
We still don't have a fleshed-out strategy how to handle translations. Although just translating an existing, English version to Swedish or French shouldn't be that complicated, the problem is maintenance. If there is say a Swedish manual which is translated to current state of the English one, and the English is then updated again... How should a translator handle this?

Using gettext
-----------------

We tend to try the gettext approach first, since we are used to it in the C++ code. However, to use gettext would need at least the translated manual(s) to be javascript code rather than static Asciidoc pages. Is this what we want?

Also, gettext works by translating separate strings. In code, these are marked with `_()`. We probably don't want to use this markup in the manual pages, though. 

A possible route would be to split the manual into sentences, use gettext to translate these (on crowdin. for example) and transform the translated manual to some javascript code displaying the translated mo-files. However, this just looks damned complicated and brittle. Thinking about it I just don't believe in it.

Using git
------------

A translator could just clone the initial version to translate by:

1. Cloning the English version
2. Making the initial translation which lives in a separate directory.
3. Making a PR against the main repo, which merges the 'sv' directory into master. 

At this point the initial translation is done. After some time, upstream/master will have an updated version of the English manual. Translator then:

1. Updates the upstream remote to reflect the new master: `git remote update upstream`
2. Checks what has been changed since last Swedish translation: `git diff master upstream/master`
3. In the 'sv' directory , updates the translation for those parts which have been changed in English
4. Makes a PR against upstream which updates the Swedish directory.
5. Pulls the local master branch so it reflects the last translated state: `git checkout master; git pull upstream master`.

The basic trick here is that the local master branch reflects the last translated state whereas  upstream/master is the latest English version. We should probably be able to use some git GUI for this, it's pretty basic.

All this should work fine.  However, it would require git knowledge beyond the level of many (most?) translators so I'm not comfortable with this approach either.

EDIT 1: Using crowdin plain text support
------------------------------------------------------

Crowdin seems to have some support for plain text, not involving gettext or similar tools.  At a glance, it looks like a direct match, but this needs to be investigated... EDIT 2: There actually seems to be asciidoc support as well: https://store.crowdin.com/adoc/. We really needs to look at this!

Thoughts? Other ideas?

**Comment by Rasbats** [comment]
@leamas : Only thinking about translation for the Basic and, more importantly, the Inland ECDIS Manual.

How about using the same strategy as OpenCPN and the plugins source code. A .pot file created from the latest en version.
@leamas : Could you look at my two scripts in the _opencpn-manuals/basic_ repo. Could they be modified to use all the pages of the en manual? Recreating all the asciidoc files in the different languages may be a big ask.

For Inland ECDIS only en, de, fr, nl are needed, I think.  This manual needs to be very simple so translations should be easier.
.po files for each page and then using **_po4a_** to recreate the asciidoc pages in the new language may not be too much work.

Already partly tested in the _translate_ directory of the _opencpn-manuals/basic_ repo.

Lots of 'shoulds' and 'mays'.

[https://store.crowdin.com/adoc/](https://store.crowdin.com/adoc/) looks interesting.

**Comment by o-charts** [comment]
Both formats Markdown (.md) and AsciiDoc (.adoc) are supported by crowdin but I am afraid this is not an option. We have experience with Markdown, not with AsciiDoc but they are quite similar. When you update a source file that contains only minor changes, most of the contents are treated as new strings, forcing translators to translate the same text over and over again. To maintain large documents, you need efficient change tracking like .pot files have. Perhaps this is a good solution for not very dynamic texts.

Send me an AsciiDoc file and I can test it.

**Reply by Rasbats** [reply]
Here is the chart.adoc I have been playing with:
[charts.adoc.pdf](https://github.com/opencpn-manuals/main/files/9092327/charts.adoc.pdf)
The .pot file for my _**po4a**_ games:
[charts.pot.pdf](https://github.com/opencpn-manuals/main/files/9092339/charts.pot.pdf)

**Reply by o-charts** [reply]
I have uploaded the adoc file. Go to https://crowdin.com/project/opencpn and select a language to translate. Then click on the file test_do_not_translate.adoc and login.When some strings are translated, let me know and I will upload the source with minor changes to see what happens.

**Reply by Rasbats** [reply]
See below.

**Comment by Rasbats** [comment]
@o-charts : French translation partly made. testing the file.

**Reply by Rasbats** [reply]
Used the partly translated .adoc and 
One problem ... I would need to make a set of screenshots for each language. Doable but time consuming
![charts](https://user-images.githubusercontent.com/2537240/178493261-bc4688c6-aa28-4e01-9f3d-98de7970333d.jpg)
.

**Reply by leamas** [reply]
The screenshots... yes, no tooling can work around this problem, for sure.

The basic question would then be if it looks good besides this. For others to judge.

On a sidenote: how would an actual, French screenshot be handled in this context? The marketing  docs says it supports it, but what does that then mean, really?

**Comment by leamas** [comment]
> How about using the same strategy as OpenCPN and the plugins source code. A .pot file created from the latest en version.

This basically does not work as I tried to describe in my initial posting. The Pot workflow ends up in .mo translations which can only be handled by code, in this case then javascript. 

Also, a meaningful  translation process requires the input to be split into separate strings. This is far from trivial, and we can certainly not use the _() markup we do in code for each and every sentence.

Let's see what Hubert comes up with, this seems like the most promising track so far to me.

**Comment by o-charts** [comment]
It seems that this has also been improved. After making some changes to the source, you are now prompted to keep or replace the translations and the rest of the strings seem to be respected:

![Screenshot from 2022-07-12 15-26-00](https://user-images.githubusercontent.com/44229499/178501954-40b93940-cee7-4b24-931d-74f9c5127f6a.png)

Maybe adding a suffix to the images could work: image1_en.png | image1_fr.png

It will not be easy to translate this format and the translator must be careful with so many labels, symbols and marks, but it could work.

You may want to ask Pavel for a manager account to be able to do more tests quickly.

Roberto

**Reply by Rasbats** [reply]
>Maybe adding a suffix to the images could work: image1_en.png | image1_fr.png

Have a look at the _inland-ecdis-docs_ repo. The en and fr directories can contain different sets of images. As long as care is taken with the naming of the images the links will work the same for each language. I will mock something up for the French translation of _charts.adoc_ tomorrow.

**Comment by Rasbats** [comment]
Updated _inland ecdis_. Latest manual is here:
[https://opencpn-manuals.github.io/inland-ecdis/manuals/en/index.html](https://opencpn-manuals.github.io/inland-ecdis/manuals/en/index.html)

We know that 'Using the manual' needs cropping. 

Experimental French translation has mainly been lost.

You may need to clear your browser web content cache to see what has changed.

**Comment by comradekingu** [comment]
https://docs.weblate.org/en/latest/formats.html?highlight=Text#translation-types-capabilities
